### DISCUSSION TITLE
What part of an element lifecycle should you fetch remote data?

### DISCUSSION THREAD

**Discussion by treeder** [lit/lit#2151] [Q&A]
I've read it could be in the constructor or on `connectedCallback`, is one better the other?

**Comment by blikblum** [comment]
Hi,

`connectedCallback` is good place to fetch data with some nuances.

Is good because occurs early in lifecycle so the first render can reflect the fetch state (loading, success, error) as opposed to using `firstUpdated` that occurs after the first render. Something i learned [creating a lit/mobx app](https://github.com/blikblum/lit-element-mobx-realworld-example-app/commit/4aab874705f2829e4f4ef0aaeeedb12afe2e8636)

One of the caveats is that `connectedCallback` can be called more than once if the element is moved in the dom tree, so if data is supposed to be fetched once some kind of flag must be used to avoid double fetch

Should be considered also if the data to be fetched depends on the value of a property to be used, e.g., as an param in an remote endpoint. Some thing like a `<person-card>` component with a `personId` property that fetches data from an endpoint 'people/:personId'.

To have a flexible component, some situations must be handled: personId is set at instantiation time, personId is NOT set at instantiation time (undefined in `connectedCallback`) and personId can be switched on the fly.

I tried to fetch both in `connectedCallback` (handle property set at instantiation) and in `updated` (handle property set lazily or changed). Soon i realized some synchronization issues when property is already changed while component was still showing old data. I [moved property check to `shouldUpdate`](https://github.com/blikblum/lit-element-mobx-realworld-example-app/commit/3c947e9e9e95bf80295cf36dad289a94c3b1e62f) that occurs before rendering. Still it's not ideal because `shouldUpdate` is not supposed to do this and is still a bit late (occurs after lit batching mechanism). In my projects, i've moved to tracking property changes in `update` that occurs at same microtask as `shouldUpdate` but feels more natural although the docs [do not recommend to override](https://lit-element.polymer-project.org/guide/lifecycle#update).

I'm still using lit-element v2. The v3 added new lifecycle that may fit better in this situation.

Regarding the question itself those are my thoughts. 

Generalizing a bit, in the process of improving the architecture of my projects, i'm trying to avoid fetch data in the component itself to simplify element implementation. I use a kind of controller responsible to track routing changes, connect to services and stores (app data). This controller pass fetched data to element and listen to element events to eventually fetch or update data. The element gets simpler and easier to be tested / demoed. Also the data is fetched earlier improving responsiveness
